File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.cpp
class Solution {
public:
    int dp[2][101][1001];
    
    int recurse(int i,int p,int k,vector<int>pr)
    {
        if(i==pr.size() || (k==0 && p))
            return 0;
        
        if(dp[p][k][i]!=-1)
            return dp[p][k][i];
        int u = pr[i];
        if(p)
            u*=-1;
        return dp[p][k][i] = max(recurse(i+1,p,k,pr) ,u+recurse(i+1,!p,k-p,pr));
            
    }
    int maxProfit(int k, vector<int>& p) {
        memset(dp,-1,sizeof(dp));
        return recurse(0,1,k,p);
    }
};

File: 188-best-time-to-buy-and-sell-stock-iv/188-best-time-to-buy-and-sell-stock-iv.java
class Solution {
    static int arr[][][];
    public int maxProfit(int k, int[] prices) {
        
        arr=new int[prices.length+1][2*k+1][2];
        for(int i[][]:arr)
            for(int f[]:i)
                Arrays.fill(f,-1);
        return transact(2*k,prices,false,0);
    }
    public static int  transact(int k,int []prices,boolean buyed,int i)
    {
       
        if(k==0||i==prices.length) return 0;
         if(arr[i][k][buyed==false?1:0]!=-1)return arr[i][k][buyed==false?1:0];
        
        if(!buyed)
        {
         
              int a= transact(k-1,prices,true,i+1)-prices[i]; 
            int b=transact(k,prices,false,i+1);
            return arr[i][k][buyed==false?1:0]=Math.max(a,b);
        }        
        else{
            // dp sell or not sell
            int a= transact(k-1,prices,false,i+1)+prices[i]; 
            int b=transact(k,prices,true,i+1);
            return arr[i][k][buyed==false?1:0]=Math.max(a,b);
        }
    
    }
}